In [1]:
#Importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
#Creating path shortcut
path = r'C:\Users\Longus\Documents\26-02-2023 Instacart Basket Analysis'

In [3]:
#Importing 'orders_products_merged.pkl'
df_ords_prods_merged = pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','orders_products_merged.pkl'))

In [5]:
#Creating a subset to avoid RAM issues
df = df_ords_prods_merged[:1000000]

In [6]:
df

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  order_number  \
0                  0             0   2539329        1             1   
1                  1             1   2398795        1             2   
2                  2             2    473747        1             3   
3                  3             3   2254736        1             4   
4                  4             4    431534        1             5   
...              ...           ...       ...      ...           ...   
999995       2731921       2731921   2836489   164627            14   
999996       2731962       2731962   1843600   164632             5   
999997       2731966       2731966    733106   164632             9   
999998       2731974       2731974   1650124   164632            17   
999999       2731979       2731979   3378258   164632            22   

        order_day_of_week  order_hour_of_day  days_since_prior_order  \
0                       2                  8                     NaN   
1                       3                  7                    15.0   
2                       3                 12                    21.0   
3                       4                  7                    29.0   
4                       4                 15                    28.0   
...                   ...                ...                     ...   
999995                  0                 15                     6.0   
999996                  1                 19                     9.0   
999997                  2                 22                    10.0   
999998                  6                 17                    13.0   
999999                  6                 10                    15.0   

        new_customer  product_id  add_to_cart_order  reordered _merge  \
0               True         196                  1          0   both   
1              False         196                  1          1   both   
2              False         196                  1          1   both   
3              False         196                  1          1   both   
4              False         196                  1          1   both   
...              ...         ...                ...        ...    ...   
999995         False       30489                  2          0   both   
999996         False       30489                  2          0   both   
999997         False       30489                  3          1   both   
999998         False       30489                  1          1   both   
999999         False       30489                  3          1   both   

        Unnamed: 0_y     product_name  aisle_id  department_id  prices  
0                195             Soda        77              7     9.0  
1                195             Soda        77              7     9.0  
2                195             Soda        77              7     9.0  
3                195             Soda        77              7     9.0  
4                195             Soda        77              7     9.0  
...              ...              ...       ...            ...     ...  
999995         30491  Original Hummus        67             20     7.5  
999996         30491  Original Hummus        67             20     7.5  
999997         30491  Original Hummus        67             20     7.5  
999998         30491  Original Hummus        67             20     7.5  
999999         30491  Original Hummus        67             20     7.5  

[1000000 rows x 18 columns]

## If- statement with user-defined functions

In [7]:
#Defining the function
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [8]:
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\Longus\AppData\Local\Temp\ipykernel_22576\2848743372.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [9]:
df['price_range'].value_counts()

Mid-range product    756450
Low-range product    243550
Name: price_range, dtype: int64

In [10]:
#Checking maximum price in the subset 'df'
df['prices'].max()

14.8

## If- stament with loc() function

In [11]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\Longus\AppData\Local\Temp\ipykernel_22576\1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [12]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [13]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [15]:
df['price_range_loc'].value_counts()

Mid-range product    756450
Low-range product    243550
Name: price_range_loc, dtype: int64

In [17]:
#Using loc() function on the 'ords_prods_merged'
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [18]:
df_ords_prods_merged.loc[(df_ords_prods_merged['prices'] <= 15) & (df_ords_prods_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [19]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [21]:
df_ords_prods_merged['price_range_loc'].value_counts(dropna = False)

Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: price_range_loc, dtype: int64

## If- statement with for-loops

In [24]:
df_ords_prods_merged['order_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: order_day_of_week, dtype: int64

In [27]:
#Creating a loop for the 'order_day_of_week' values
result = []

for value in df_ords_prods_merged["order_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [28]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [29]:
df_ords_prods_merged['busiest_day'] = result

In [31]:
df_ords_prods_merged['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

## Task Exercise

### Creating new column 'Busiest days'

In [32]:
df_ords_prods_merged['order_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: order_day_of_week, dtype: int64

#### Busiest days are 0 and 1 (Saturday and Sunday), least busy days are 3 and 4 (Tuesday and Wednesday)

In [40]:
#Creating a loop for the 'order_day_of_week' values
days = []

for value in df_ords_prods_merged['order_day_of_week']:
  if value == 0 or value == 1:
    days.append('Busiest days')
  elif value == 3 or value == 4:
    days.append('Slowest days')
  else:
    days.append('Regularly busy')

In [41]:
#Creating the column 'busiest_days' and adding to the dataframe
df_ords_prods_merged['busiest_days'] = days

In [43]:
#Checking value count
df_ords_prods_merged['busiest_days'].value_counts(dropna = False)

Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: busiest_days, dtype: int64

In [37]:
df_ords_prods_merged.shape

(32404859, 20)

#### Sum of counts of values equal to total number of rows in the dataframe 'df_ords_prods_merged'

### Creating new column 'busiest_period_of_day'

In [62]:
#Checking the frequency of the values in 'order_hour_of_day'
count_of_orders = df_ords_prods_merged['order_hour_of_day'].value_counts(dropna = False)

In [63]:
count_of_orders

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

In [75]:
#Searching for percentile of the count of values
list = list(df_ords_prods_merged['order_hour_of_day'].value_counts(dropna = False).quantile([0.75, 0.5, 0.25]))

TypeError: 'list' object is not callable

In [74]:
list

[2556034.5, 1117230.5, 272562.0]

##### Based on the percentiles, we can divide the hours into 3 groups:
##### Group 1 (Fewest orders - Under 272562 orders): 3, 4, 2, 5, 1, 0
##### Group 2 (Average orders - Between 272562 and 2556034.5): 6, 23, 22, 21, 7, 20,19,18,8,17,9,16
##### Group 3 (Most orders - Over 2556034.5): 12,13,15,14,11,10

In [77]:
#Creating a loop for 'order_hour_of_day' values
hours = []

for value in df_ords_prods_merged['order_hour_of_day']:
  if value in [10, 11, 14, 15, 13, 12]:
    hours.append('Most orders')
  elif value in [0, 1, 5, 2, 4, 3]:
    hours.append('Fewest orders')
  else:
    hours.append('Average orders')

In [78]:
#Adding the column 'busiest_period_of_day' to the dataframe
df_ords_prods_merged['busiest_period_of_day'] = hours

In [79]:
df_ords_prods_merged.head()

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  order_number  \
0             0             0   2539329        1             1   
1             1             1   2398795        1             2   
2             2             2    473747        1             3   
3             3             3   2254736        1             4   
4             4             4    431534        1             5   

   order_day_of_week  order_hour_of_day  days_since_prior_order  new_customer  \
0                  2                  8                     NaN          True   
1                  3                  7                    15.0         False   
2                  3                 12                    21.0         False   
3                  4                  7                    29.0         False   
4                  4                 15                    28.0         False   

   product_id  ...  _merge  Unnamed: 0_y product_name  aisle_id department_id  \
0         196  ...    both           195         Soda        77             7   
1         196  ...    both           195         Soda        77             7   
2         196  ...    both           195         Soda        77             7   
3         196  ...    both           195         Soda        77             7   
4         196  ...    both           195         Soda        77             7   

   prices    price_range_loc     busiest_day    busiest_days  \
0     9.0  Mid-range product  Regularly busy  Regularly busy   
1     9.0  Mid-range product  Regularly busy    Slowest days   
2     9.0  Mid-range product  Regularly busy    Slowest days   
3     9.0  Mid-range product      Least busy    Slowest days   
4     9.0  Mid-range product      Least busy    Slowest days   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2           Most orders  
3        Average orders  
4           Most orders  

[5 rows x 22 columns]

In [80]:
df_ords_prods_merged['busiest_period_of_day'].value_counts(dropna = False)

Most orders       16128666
Average orders    15679865
Fewest orders       596328
Name: busiest_period_of_day, dtype: int64

In [81]:
df_ords_prods_merged.shape

(32404859, 22)

#### The sum of counts of 'busiest_period_of_day' equals to total number of rows in dataframe 'df_ords_prods_merged'

In [82]:
#Exporting the dataframe in .pkl format
df_ords_prods_merged.to_pickle(os.path.join(path,'02 Data','Prepared Data','orders_products_merged_updated.pkl'))